src/main/java/com/google/sps/data/FirestoreDB.java

```diff
@@ -0,0 +1,140 @@
+import com.google.cloud.firestore.DocumentSnapshot;
+import com.google.cloud.firestore.FieldValue;
+import com.google.cloud.firestore.Firestore;
+import com.google.cloud.firestore.Query;
+import com.google.firebase.cloud.FirestoreClient;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+public class FirestoreDB implements DBInterface {
+    public String getRecipeContent(String Id) {
+        return null;
+    }
+
+    public RecipeMetadata getRecipeMetadata(String Id) {
+        return null;
+    }
+    public String addRecipe(RecipeMetadata newRecipe, String newContent) {
+        return null;
+    }
+    public void deleteRecipe(String Id) {
+        return;
+    }
+
+    public void editRecipeTitleContent(String editedTitle, String editedContent) {
+        return;
+    }
+
+    public Iterable<RecipeMetadata> getAllRecipes() {
+        return null;
+    }
+    public Iterable<Comment> getAllCommentsInRecipe(String recipeId) {
+        return null;
+    }
+    public Iterable<Tag> getAllTags() {
+        return null;
+    }
+
+    /** Returns a User object from userId. */
+    public User getUser(String userId) {
+      DocumentReference userRef = DBUtils.user(userId);
+      DocumentSnapshot user = DBUtils.blockOnFuture(userRef.get());
+      return user.toObject(User.class);
+    }
+   
+    /** Adds a User to the db based on an input userId. */
+    public String addUser(String userId) {
+      DocumentReference newUserRef = DBUtils.users().document(userId);
+      User newUser = new User(userId);
+      DBUtils.blockOnFuture(newUserRef.set(newUser));  // .set() returns an ApiFuture
+      return userId;
+    }
+    
+    /** Delete a User from the db based on an input userId. */
+    public void deleteUser(String userId) {
+      DBUtils.blockOnFuture(DBUtils.user(userId).delete());
+    }
+    
+    /** Add recipe id: true as a singleton map to user's created recipes field. */
+    public void addRecipeIdToCreated(String userId, String recipeId) {
+      DocumentReference userRef = DBUtils.user(userId);
+      String nestedPropertyName = DBUtils.getNestedPropertyName(User.CREATED_RECIPES_KEY, recipeId);
+      ApiFuture addToCreatedPostsFuture =
+          userRef.update(Collections.singletonMap(nestedPropertyName, true));
+      DBUtils.blockOnFuture(addToCreatedPostsFuture);
+    }
+    
+    /** Remove a recipe from a user's created (when recipe is deleted). */
+    public void removeRecipeIdFromCreated(String userId, String recipeId) {
+      DocumentReference userRef = DBUtils.user(userId);
+      String nestedPropertyName = DBUtils.getNestedPropertyName(User.CREATED_RECIPES_KEY, recipeId);
+      Map<String, Object> update = new HashMap<>();
+      update.put(nestedPropertyName, FieldValue.delete());
+      DBUtils.blockOnFuture(userRef.update(update));
```

Comment:
This can all be super condensed using `userRef.update(nestedPropertyName , FieldValue.delete())`

Reply:
Additionally, you might want to make `removeRecipeFromUserCollection(String collection)` a private function for use in this and similar functions.

Reply:
This is definitely not the best place for this but I’m sending this message from my phone, I don’t have WiFi at the moment so if I’m not in the meeting by 10:05 please start without me

Reply:
We're delaying the meeting till 2:30.